scala_proto/private/scala_proto_aspect.bzl

```diff
@@ -122,27 +133,61 @@ def _compile_sources(ctx, toolchain, proto, src_jars, deps):
         runtime_deps = compile_deps,
     )
 
-####
-# This is applied to the DAG of proto_librarys reachable from a deps
-# or a scalapb_scala_library. Each proto_library will be one scalapb
-# invocation assuming it has some sources.
-def _scalapb_aspect_impl(target, ctx):
+def _phase_proto_provider(ctx, p):
+    target = p.target
+    return target[ProtoInfo]
+
+def _phase_deps(ctx, p):
+    return [d[ScalaProtoAspectInfo].java_info for d in ctx.rule.attr.deps]
+
+def _phase_generate_and_compile(ctx, p):
```

Comment:
Would it be possible to split this phase into two without `and`?

Reply:
It's possible, I didn't not do it to avoid multiple checks accross phases for this conditional decision:
```starlark
if proto.direct_sources and _code_should_be_generated(ctx, toolchain):
    src_jars = _generate_sources(ctx, toolchain, proto)
    java_info = _compile_sources(ctx, toolchain, proto, src_jars, deps, stamp_label)
    return java_info
else:
    # this target is only an aggregation target
    return java_common.merge(deps)
```
Do you think it's worth splitting it?

Reply:
It's not a blocker for sure. It just feels unbalanced. Some phases are fine grained and this one is coarse grained doing multiple things. Maybe phases are missing some "short-circuit" functionality. So let's keep it that way if it's simpler.